### PR TITLE
Add test workflow

### PR DESCRIPTION
**Pull request by gnodar01**
Setup a github action to run tests

### CODE REVIEW

**Comment by gnodar01** [comment]
Should be squashed on merge.

**Comment by bethac07** [comment]
>Right now I have tests running, with caching, just on python 3.8, and macos-11 (i.e.[macos-latest](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources)). macOS 10.15, which we currently have release.yml set to use, has been https://github.com/actions/virtual-environments/issues/5583 and will be completely phased out 8/30/22.

Makes sense to do that for tests, but builds we often want to keep the oldest one we can - these are our Google Analytics of visitors to the CellProfiler site for the last 6 weeks, filtered to just the Mac users, but this seems highly suspicious and likely to be wrong to me. But since if we build on 10.15, 11.X users CAN use, but the reverse isn't true, I'd err on the side of building 4.2.2 on 10.15. What do you think? 
image

**Comment by bethac07** [comment]
>I could add windows-latest (windows-2022) and ubuntu-latest (ubuntu-20.04) to the os matrix, but 1) I'll need to get mysql installed and running on each one (if there are already instructions somewhere on this, please point me to them), and 2) not sure if it's even worth it, given the amount of time it takes on just macos.

I would think that the Windows build instructions for the app, and the ubuntu installation instructions on the GH wiki, would probably take care of most of the pain of this. I think matrixed stuff also typically all builds simultaneously, rather than sequentially (at least up to 4-5 runners), so it wouldn't necessarily slow the testing unless one was dramatically slower than Mac.

With all that being said, I would say the priority for doing this would be low-medium for Windows (there ARE some Windows-only tests) and low-low for Linux. If you forced me to pick between matrixing these and Python versions, I'd definitely pick Python versions.

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      - name: Install CP
+        run: |
+          pip install -e ~/python-javabridge
+          pip install -e ~/python-bioformats
+          pip install -e ~/centrosome
+          pip install -e ~/core
+          pip install -e .[test]
+      - name: Test
+        run: |
+          # mysql.server status
```
Comment (bethac07) [inline .github/workflows/test.yml:70]: Any reason to keep these?

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+          sudo apt-get install -y default-libmysqlclient-dev libnotify-dev libsdl2-dev
+      - name: Mac - setup mysql
+        env:
+          LC_ALL: "en_US.UTF-8"
+        if: startsWith(matrix.os, 'macos')
+        run: |
+          brew install mysql
+          mysql.server start
+          mysql --host=$CP_MYSQL_TEST_HOST --user=$CP_MYSQL_TEST_USER --execute="CREATE DATABASE $CP_MYSQL_TEST_DB;" --skip-password
+          # mysql --host=$CP_MYSQL_TEST_HOST --user=$CP_MYSQL_TEST_USER --execute="CREATE DATABASE $CP_MYSQL_TEST_DB;" --password=$CP_MYSQL_TEST_PASSWORD
```
Comment (bethac07) [inline .github/workflows/test.yml:60]: Any reason to keep this?

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        run: |
+          pip install pyinstaller
+          pip install --upgrade pip setuptools wheel
+          pip install numpy>=1.20.1
+          git clone https://github.com/CellProfiler/python-javabridge.git ~/python-javabridge
+          git clone https://github.com/CellProfiler/python-bioformats.git ~/python-bioformats
+          git clone https://github.com/CellProfiler/centrosome.git ~/centrosome
+          git clone https://github.com/CellProfiler/core.git ~/core
+          # git clone https://github.com/CellProfiler/CellProfiler.git ~/cellprofiler
+      - name: Linux - install mysql and build tools
```
Comment (bethac07) [inline .github/workflows/test.yml:45]: If we decide against adding linux for now, do we wan to clean this up?

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      - name: Initial installs
+        run: |
+          pip install pyinstaller
+          pip install --upgrade pip setuptools wheel
+          pip install numpy>=1.20.1
+          git clone https://github.com/CellProfiler/python-javabridge.git ~/python-javabridge
+          git clone https://github.com/CellProfiler/python-bioformats.git ~/python-bioformats
+          git clone https://github.com/CellProfiler/centrosome.git ~/centrosome
+          git clone https://github.com/CellProfiler/core.git ~/core
+          # git clone https://github.com/CellProfiler/CellProfiler.git ~/cellprofiler
```
Comment (bethac07) [inline .github/workflows/test.yml:44]: Should be able to remove this now, yes? woohoo!

**Review by bethac07 — APPROVED** [review]
A couple minor things you may want to clean up (or not, if they're notes to future you!), but otherwise, yay, thanks for getting this working!

File: .github/workflows/test.yml (hunk 4 above)
Comment (bethac07) [reply .github/workflows/test.yml:44]: Never mind, I lost track of what repo I was in, I am mixed up lol

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+          LC_ALL: "en_US.UTF-8"
+        if: startsWith(matrix.os, 'macos')
+        run: |
+          brew install mysql
+          mysql.server start
+          mysql --host=$CP_MYSQL_TEST_HOST --user=$CP_MYSQL_TEST_USER --execute="CREATE DATABASE $CP_MYSQL_TEST_DB;" --skip-password
+          # mysql --host=$CP_MYSQL_TEST_HOST --user=$CP_MYSQL_TEST_USER --execute="CREATE DATABASE $CP_MYSQL_TEST_DB;" --password=$CP_MYSQL_TEST_PASSWORD
+      - name: Install CP
+        run: |
+          pip install -e ~/python-javabridge
```
Comment (bethac07) [inline .github/workflows/test.yml:63]: I don't think it's a _bad_ idea to build all of these from source, but I think if we wanted to shave some time then using PyPI wheels for most of these should typically work.

**Comment by gnodar01** [comment]
> I'd err on the side of building 4.2.2 on 10.15

I'm in general agreement, but on the last run I tried with `macos-10.15`, it was [automatically terminated](https://github.com/CellProfiler/CellProfiler/actions/runs/2712961459) because of the deprecation, stating:

> This is a scheduled macOS-10.15 brownout.

So we have to make sure we run the 4.2.2 release workflow outside of the brownout periods (posted in that [same deprecation issue](https://github.com/actions/virtual-environments/issues/5583)).

File: .github/workflows/test.yml (hunk 1 above)
Comment (gnodar01) [reply .github/workflows/test.yml:70]: Nope, will remove

File: .github/workflows/test.yml (hunk 2 above)
Comment (gnodar01) [reply .github/workflows/test.yml:60]: No, at least for now. I had some trouble around [this line](https://github.com/CellProfiler/CellProfiler/blob/master/tests/modules/test_exporttodatabase.py#L114), but I decided it was outside of the scope of this PR, so I'll remove it for now.

File: .github/workflows/test.yml (hunk 3 above)
Comment (gnodar01) [reply .github/workflows/test.yml:45]: Will do.

File: .github/workflows/test.yml (hunk 5 above)
Comment (gnodar01) [reply .github/workflows/test.yml:63]: Okay. I'll keep core from source, since we always want to have the latest, and will remove the others.

**Comment by bethac07** [comment]
Shall we?

**Comment by gnodar01** [comment]
Note for future self, and others:

For some reason the "Cache pip" step, [this run](https://github.com/CellProfiler/CellProfiler/runs/7459555397?check_suite_focus=true) reports:

> Cache not found for input keys: macOS-pip-a6443ba1eee7b09cafd18f4b377dac37bec6ff0e3c4f60e9be4b42dd723a816d

Even though only a few minutes before [the previous successful run](https://github.com/CellProfiler/CellProfiler/runs/7458784769?check_suite_focus=true) reported:

> Cache restored from key: macOS-pip-a6443ba1eee7b09cafd18f4b377dac37bec6ff0e3c4f60e9be4b42dd723a816d

and in the post cache step:

> Cache hit occurred on the primary key macOS-pip-a6443ba1eee7b09cafd18f4b377dac37bec6ff0e3c4f60e9be4b42dd723a816d, not saving cache.

The GH CLI is also reporting a cache exists with that key, so it seemed a bit weird that there was a cache miss. I'm reasonably certain that it's because I was not including the python version in the key generation, so the python 3.8 cache save was overwriting the python 3.9 cache contents, and vice versa.

**Comment by gnodar01** [comment]
> If you forced me to pick between matrixing these and Python versions, I'd definitely pick Python versions.

One minor thing to note about this, as long as we're using `wxPython==4.1.0`, cache misses are going to lead to slow installs on python 3.9, mac and windows included, since only `4.1.1` has a python 3.9 wheel on PyPI. 

Not really a big deal, but one option is to host our own macos 11 + python 3.9 + wxPython 4.1.0 wheel...

**Pull request merged**